### DISCUSSION TITLE
Space Mouse support

### DISCUSSION THREAD

**Discussion by forrestjgq** [isl-org/Open3D#4585] [Ideas]
Space mouse designed by [3Dconnexion](http://www.3dconnexion.com/) is quite useful for 3D navigating. I've implemented that in my own branch by using a thirdparty library [libspnav](https://github.com/FreeSpacenav/libspnav) for both gui SceneWidget and Visualizer. It works fine.

I wonder if it's necessary to submit a PR to support this feature?

**Comment by forrestjgq** [comment]
Code reference of my own branch:
https://github.com/forrestjgq/Open3D/pull/10
https://github.com/forrestjgq/Open3D/pull/11

**Reply by olagt** [reply]
Hi @forrestjgq I tried your fork branch "lego15" , with the latest set spacenavd, libspnav,spnavcfg . It looks like  my spacemouse itself is working correctly, however  open3d doesn't react on any movement  :(  I tried Visualizer from examples and ViewGeometry. I also compiled my program against your fork,  based on denseSLAMGUI where widget3d is used for 3d display , also with no avail :(
 Any ideas or clues what I'm doing wrong ?  

  this->widget3d_->GetScene()->GetScene();
  scene->AddGeometry("points", *pointcloud_ptr2 ,mat);

**Reply by forrestjgq** [reply]
@olagt Have you started spacenavd?  
`sudo /etc/init.d/spacenavd start`

you may:
1. check if libspnav examples to see if service daemon is started correctly, it's in libspnav/examples/cube
2. check the cmake printing, there should be `Using installed third-party library 3rdparty_spnav ` printed, the open3d should be cleaned and built **after** spacenavd/libspnav is installed
3. `lego15` branch use socket to connect to spnavd, once connect fails at starting, it won't retry. You could add some logs here: `https://github.com/forrestjgq/Open3D/blob/lego15/cpp/open3d/visualization/utility/SpaceMouse.cpp#L52`, build and try again
4. I always use Open3D program to load any ply, set in **arcball** mode to try. `make Open3dViewer && ./bin/Open3D/Open3D`

**Reply by olagt** [reply]
@forrestjgq Thank you for great guidance ! I followed the steps and my space mouse is working great.

**Comment by meyerjo** [comment]
Would be nice to see this feature in open3d

**Comment by olagt** [comment]
Hi @forrestjgq , do you plan to make PR to current 0.16 of your great support for SpaceMouse ?
I really miss this functionally in Open3D  :( 
Kind regards

Ola
